### PR TITLE
add smart nullable

### PR DESCRIPTION
**Pull request by simonkimi**
在开启空安全后, value为null的字段将会默认勾选可空
可以直接解析List<T>类型的Json了, 字段为取并集
对部分有setter但没有使用setter进行修改

### CODE REVIEW

**Comment by zmtzawqlp** [comment]
。。。你勾上。右边最上面的可空。不就可以了。。。。

**Comment by zmtzawqlp** [comment]
it seems that smart nullable is better than auto nullable

**Comment by zmtzawqlp** [comment]
run dartfmt -w .  before you push pr

**Comment by simonkimi** [comment]
done

**Comment by simonkimi** [comment]
修改说明:
1. `DartObject` 的 `_jObject`由原先的`Map<String, dynamic>`修改为`Map<String, Tuple3<dynamic, DartType, bool>>`, Tuple3中分别为 数据/类型/是否可空, 记录类型可在`DartObject.merge`中对类型进行对比并去除`Null`, 此举是为解决形如`[{"foo": null}, {"foo": 123}]`的foo类型被解析为Object?
2. 对于`List<dynamic>`的Json解析被改为为List添加一个头ArrayRoot, 将List转换为`Map<List<dynamic>>`进行解析, 随后去除ArrayRoot节点, 将depth整体减一. 此举可对列表查空的逻辑进行复用.

File: Flutter/json_to_dart/lib/models/config.dart [hunk 1]
```diff
@@ -90,6 +90,17 @@ class ConfigSetting extends Setting<ConfigSetting> with ChangeNotifier {
       notifyListeners();
     }
   }
+
+  bool _smartNullable = false;
+  @HiveField(13)
+  bool get smartNullable => _smartNullable;
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/models/config.dart:96]: bool get smartNullable => _smartNullable??false;

hive issue https://github.com/hivedb/hive/issues/625

Comment (simonkimi) [reply Flutter/json_to_dart/lib/models/config.dart:96]: I think it's unnecessary, the exception is thrown from config.g.dart `..smartNullable = fields[13] as bool`, `type 'Null' is not a subtype of type 'bool' in type cast`,  I directly modified config.g.dart to solve this problem temporarily, but I still suggest adding the version number to the hive.

File: Flutter/json_to_dart/lib/pages/json_tree_header.dart [hunk 2]
```diff
@@ -17,9 +19,9 @@ class JsonTreeHeader extends StatefulWidget {
 class _JsonTreeHeaderState extends State<JsonTreeHeader> {
   @override
   Widget build(BuildContext context) {
-    // final JsonToDartController controller =
-    //     Provider.of<JsonToDartController>(context);
-    final AppLocalizations appLocalizations = AppLocalizations.of(context);
+    final JsonToDartController controller =
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/pages/json_tree_header.dart:22]: rock back this

File: Flutter/json_to_dart/lib/pages/setting.dart [hunk 3]
```diff
@@ -71,6 +72,7 @@ class _SettingWidgetState extends State<SettingWidget> {
             onChanged: (Locale? value) {
               setState(() {
                 ConfigSetting().locale = value!;
+                ConfigSetting().save();
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/pages/setting.dart:75]: rock back

File: Flutter/json_to_dart/lib/pages/json_tree_item.dart [hunk 4]
```diff
@@ -338,7 +343,7 @@ class _NullableCheckBoxState extends State<NullableCheckBox> {
       value: widget.property.nullable,
       onChanged: (bool value) {
         setState(() {
-          widget.property.nullable = value;
+          widget.property.updateNullable(value);
```
[the remark this answers is not in the document]
Comment (zmtzawqlp) [reply Flutter/json_to_dart/lib/pages/json_tree_item.dart:346]: roll back

File: Flutter/json_to_dart/lib/pages/json_tree_item.dart [hunk 5]
```diff
@@ -215,7 +215,7 @@ class _JsonTreeItemState extends State<JsonTreeItem> {
                   ),
           );
         } else {
-          property.nullable = true;
+          property.updateNullable(true);
```
[the remark this answers is not in the document]
Comment (zmtzawqlp) [reply Flutter/json_to_dart/lib/pages/json_tree_item.dart:218]: roll back

File: Flutter/json_to_dart/lib/models/config.g.dart [hunk 6]
```diff
@@ -26,17 +26,18 @@ class ConfigSettingAdapter extends TypeAdapter<ConfigSetting> {
       ..traverseArrayCount = fields[6] as int
       ..propertyNamingConventionsType =
           fields[7] as PropertyNamingConventionsType
-      ..propertyAccessorType = fields[8] as PropertyAccessorType
       ..propertyNameSortingType = fields[9] as PropertyNameSortingType
+      ..propertyAccessorType = fields[8] as PropertyAccessorType
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/models/config.g.dart:30]: change code if it's necessary

File: Flutter/json_to_dart/lib/models/config.g.dart [hunk 7]
```diff
@@ -53,16 +54,18 @@ class ConfigSettingAdapter extends TypeAdapter<ConfigSetting> {
       ..write(obj.traverseArrayCount)
       ..writeByte(7)
       ..write(obj.propertyNamingConventionsType)
-      ..writeByte(8)
-      ..write(obj.propertyAccessorType)
       ..writeByte(9)
       ..write(obj.propertyNameSortingType)
+      ..writeByte(8)
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/models/config.g.dart:59]: change  if it's necessary

File: Flutter/json_to_dart/lib/models/dart_object.dart [hunk 8]
```diff
@@ -19,19 +21,35 @@ class DartObject extends DartProperty {
     MapEntry<String, dynamic>? keyValuePair,
     required int depth,
     DartObject? source,
+    required bool nullable,
   }) : super(
             uid: source?.uid ?? uid!,
             keyValuePair: source?.keyValuePair ?? keyValuePair!,
-            depth: source?.depth ?? depth) {
+            depth: source?.depth ?? depth,
+            nullable: nullable) {
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/models/dart_object.dart:29]: nullable: nullable,)

better

File: Flutter/json_to_dart/lib/pages/setting.dart [hunk 9]
```diff
@@ -232,6 +235,21 @@ class _MoreSettingState extends State<MoreSetting> {
               //controller.updateNullable(true);
               ConfigSetting().nullsafety = value;
               ConfigSetting().nullable = true;
+              if (!value) {
+                ConfigSetting().smartNullable = false;
+              }
+            });
+          },
+        ),
+        StCheckBox(
```
Comment (zmtzawqlp) [inline Flutter/json_to_dart/lib/pages/setting.dart:244]: if(ConfigSetting().nullsafety)
   xxxx

**Comment by zmtzawqlp** [comment]
LGTM, thanks for your contribution

**Pull request merged**